File: 3548-find-the-count-of-good-integers/find-the-count-of-good-integers.go
func factorial(n int64) *big.Int {
	res := big.NewInt(1);

	for i := int64(2); i <= n; i++ {
		res.Mul(res, big.NewInt(i));
	}

	return res;
}

func generatePalindromes(n, k int) []string {
	var result []string;
	var dfs func(i int, curr []byte);

	dfs = func(i int, curr []byte) {
		if i >= (n+1)/2 {
			if curr[0] == '0' {
				return;
			}

			num := new(big.Int);
			num.SetString(string(curr), 10);

			if new(big.Int).Mod(num, big.NewInt(int64(k))).Cmp(big.NewInt(0)) == 0 {
				result = append(result, string(curr));
			}

			return;
		}

		for c := byte('0'); c <= '9'; c++ {
			if i == 0 && c == '0' {
				continue;
			}

			curr[i], curr[n-1-i] = c, c;
			dfs(i+1, curr);
		}
	}

	dfs(0, make([]byte, n));
	return result;
}

func countGoodIntegers(n int, k int) int64 {
	validPalindromes := generatePalindromes(n, k);

	uniquePatterns := make(map[string]bool);

	for _, pal := range validPalindromes {
		freq := make([]int, 10);

		for _, ch := range pal {
			freq[ch-'0']++;
		}

		builder := strings.Builder{};

		for _, f := range freq {
			builder.WriteByte(byte('a' + f));
		}

		uniquePatterns[builder.String()] = true;
	}

	totalPerm := factorial(int64(n));
	totalValid := big.NewInt(0);

	for pattern := range uniquePatterns {
		perm := new(big.Int).Set(totalPerm);

		for i := 0; i < len(pattern); i++ {
			perm.Div(perm, factorial(int64(pattern[i]-'a')));
		}

		if pattern[0] != 'a' {
			zeroCount := int(pattern[0] - 'a');

			if zeroCount > 0 {
				invalid := factorial(int64(n - 1));

				for i := 1; i < len(pattern); i++ {
					invalid.Div(invalid, factorial(int64(pattern[i]-'a')));
				}

				invalid.Div(invalid, factorial(int64(zeroCount-1)));
				perm.Sub(perm, invalid);
			}
		}

		totalValid.Add(totalValid, perm);
	}

	return totalValid.Int64();
}

File: 3548-find-the-count-of-good-integers/find-the-count-of-good-integers.java
class Solution {
    public long factorial(int num) {
        long result = 1;
        for (int i = 1; i <= num; i++) result *= i;

        return result;
    }

    public void generatePalindromes(char[] currentNumber, int index, List<String> validPalindromes, int k) {
        if (index >= (currentNumber.length + 1) / 2) {
            String palindrome = new String(currentNumber);
            if (palindrome.charAt(0) != '0' && Long.parseLong(palindrome) % k == 0) {
                validPalindromes.add(palindrome);
            }

            return;
        }

        for (char c = '0'; c <= '9'; c++) {
            if (index == 0 && c == '0') continue;
            currentNumber[index] = c;
            currentNumber[currentNumber.length - 1 - index] = c;
            generatePalindromes(currentNumber, index + 1, validPalindromes, k);
        }
    }

    public long countGoodIntegers(int n, int k) {
        List<String> validPalindromes = new ArrayList<>();
        char[] basePalindrome = new char[n];
        Arrays.fill(basePalindrome, '0');
        generatePalindromes(basePalindrome, 0, validPalindromes, k);

        Set<String> uniquePatterns = new HashSet<>();

        for (String palindrome : validPalindromes) {
            int[] freq = new int[10];
            for (char ch : palindrome.toCharArray()) freq[ch - '0']++;
            StringBuilder pattern = new StringBuilder();
            for (int f : freq) pattern.append((char) ('a' + f));
            uniquePatterns.add(pattern.toString());
        }

        long totalPermutations = factorial(n);
        long totalValidPermutations = 0;

        for (String pattern : uniquePatterns) {
            long permutations = totalPermutations;
            for (char ch : pattern.toCharArray()) {
                permutations /= factorial(ch - 'a');
            }

            if (pattern.charAt(0) != 'a') {
                int zeroCount = pattern.charAt(0) - 'a';
                if (zeroCount > 0) {
                    long invalidPerms = factorial(n - 1);
                    for (int j = 1; j < pattern.length(); j++) {
                        invalidPerms /= factorial(pattern.charAt(j) - 'a');
                    }
                    invalidPerms /= factorial(zeroCount - 1);
                    permutations -= invalidPerms;
                }
            }

            totalValidPermutations += permutations;
        }

        return totalValidPermutations;
    }
}

// Sinh ra tất cả các số đối xứng (palindrome) có n chữ số và chia hết cho k.
// Với mỗi số palindrome, tạo ra một chuỗi pattern tần suất chữ số (ví dụ aabc với a=2, b=1, c=1).
// Với mỗi pattern, tính số hoán vị khác nhau của n chữ số tương ứng pattern đó.
// Trừ đi các hoán vị bắt đầu bằng số 0.
// Cộng tất cả các hoán vị hợp lệ lại.